keps/sig-node/20200303-shared-pod-limits.md

```diff
@@ -0,0 +1,327 @@
+
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+The `Burstable` QoS level is used for any Pod that defines resource limits for at least one container, and doesn't define the same value for both the `request` and `limit` sections. In these Pods, a container which doesn't define any limit for a resource will effectively not be constrained in any way for that resource. Defining a limit for all of the containers in the pod has the implication that it is not possible to define a pod where the resources are controlled on the pod level. 
+
+This KEP proposes a way to define constraints on the Pod level, to allow for a smoother resource allocation strategy for `Burstable` pods.
+
+
+## Motivation
+
+
+Some workloads are deployed as pods which are comprised of multiple sidecar containers which are strongly coupled in terms of their task. Such containers communicate either across a shared filesystem, or the localhost network, and orchestrate some common task. 
+
+For example, consider the following Pod with the following structure:
+
+<pre>
+  pod
+   |
+   +-- container1 (main task)
+   |
+   +-- container2 (second level task)
+   |
+   +-- container3 (log handler)
+   |
+   +-- container4 (mesh sidecar)
+</pre>
+
+In some cases deploying a single container with all the tasks is not optimal and not always possible. Kubernetes is not aware of these tasks, and doesn't monitor them for failure, and is not able to manage the resources (cgroups) for each of them. By separating the different tasks to their own containers, the application is able to leverage Kubernetes to monitor the tasks. The current `Burstable` QoS implementation requires the pod to either not limit each individual container at all, or micro-manage the resources allocated to each and every container.
+
+This proposal suggests a middle ground, and suggests a way to make it possible to describe to Kubernetes how to limit the resource consumption of multiple containers in the pod at once, on the Pod level, instead of trying to micro-manage the resource limits on the container level. For workloads where the work performed is burstable, this proposal would make it easier to allow the low-level mechanisms available in the underlying operating-system to manage the resources required for the task.
+
+### Goals
+
+This proposal aims to:
+
+*  Allow a `Burstable` pod to define that memory and CPU resource limits should be set on the level of the Pod.
+*  Prevent the developer from having to micro-manage the memory and CPU resource assignments for different containers in the same pod.
```

Comment:
@derekwaynecarr do we think this motivation still holds after "In-place updates for pod-resources" lands?

Reply:
@dashpole please consider that some burstable workloads are such that most of the time they need low amounts of a resource, but then during a burst of activity they need a lot of it, and fast. There is no time in these scenarios for a developer or even a VPA to get involved and scale the resource. At least for memory, by the time someone or something can react the process could already have been killed by the OOM killer. 

At least for my use-case (a web based IDE deployed as a pod) having the ability to change the requested limits after the fact is not good enough. My pods have a defined cap on the amount of allowed resources. Even if Kubernetes allowed modifying the per-container limits dynamically, it's impossible to have a control loop that would be as responsive as the actual cgroup implementation provided by the Linux kernel.

Reply:
My question was related to micro-management motivation bullet.  I'll certainly grant that VPA isn't a silver bullet for all resource-management problems, but if it is made to be able to adjust resources based on usage with some frequency, it should be able to eliminate the need to adjust container settings.

IMO, your motivation seems to be around bounding the resource usage of the pod when containers don't have predictable resource usage.  If you were to set container limits based on peak usage, the pod limit would be enormous.

If you framed the proposal in terms of an additional static limit on the pod _in addition to_ container limits, I think it makes more sense.  I think you are getting a lot of pushback on the idea of _removing_ container limits, where your real motivation is to make the pod-level cgroup lower than it would otherwise be if we summed container-level limits.

Reply:
I was only addressing the micro-management motivation point.  I completely agree that VPA is not a silver bullet that solves all resource management problems.

IMO, your motivation is primarily about bounding pod usage when container usage is unpredictable.  If you sized container limits with the peak container usage, the overall limit would be enormous.  You would like to decrease the amount by which your pod is overcommitted to make the system more stable.

I think you are running into pushback over the idea that this should be a substitute for container-level limits.  IMO this should be framed simply as an additional constraint placed on pods.  Users should still set container limits, but they can now additionally constrain the overall usage of the pod when container limits are sized for peak.  It still has some poor interactions with in-place updates of container resources, and adds complexity to setting pod/container resources, but at least won't raise the sorts of questions about running without limits you got at sig-node today.

Reply:
The problem I see is that if a container is run with memory limits, then a burstable workload will reach OOM since this limit is enforced by the cgroup.

I still want a limit, but I really want it to be set on the upper cgroup in the hierarchy, not on the leaf cgroup.

The issues raised in the sig-node meeting were valid, I will research this some more to come up with good answers.